File: sce.cpp

#include <boost/random.hpp>
#include <algorithm>
#include <iostream>
#include <math.h>

#include "constants.hpp"
#include "droplet.hpp"
#include "util.hpp"
#include "collisions.hpp"

using namespace std;
using namespace constants;

const bool DEBUG = false;

int main() {

    double delta_V = 1e6; // Cell volume, m^3
    double t_c = 1.0; // timestep, seconds
    const resolution n_part = lo;

    // CASE SETTINGS
    int t_end = 3601;
    int plot_dt = 1200;
    double n_0 = pow(2., 23);
    double R_0 = 30.531e-6;
    double X_0 = (4. * M_PI / 3.) * (R_0 * R_0 * R_0);
    double M_0 = X_0 * RHO_WATER;
    double m_tot_ana = 1.0;

    /* ************************************************ */

    // Create a droplet array
    Droplet droplets[n_part];

    // Pre-compute the super-droplet multiplicity
    double total_droplets = delta_V * n_0;
    long xi_i = long(floor(total_droplets / n_part));

    // Using an exponential distribution, sample droplet masses and initialize
    // the droplets.
    boost::random::mt19937 rng;
    boost::random::exponential_distribution<> dist(1. / X_0);
    for (int i = 0; i < n_part; i++) {
        double x = dist(rng);
        double r = pow(x * 3. / M_PI / 4., 1. / 3.);
        droplets[i] = Droplet(xi_i, pow(r, 3.));
//        cout << droplets[i] << "\n";
    }
    sort(droplets, droplets + n_part, smaller); // Sort using a comparator function

    cout << "GRID SETUP" << endl;
    cout << "   radii: " << droplets[0].get_radius() << " - "
    << droplets[n_part - 1].get_radius() << " m" << endl;
    cout << "  volume: " << droplets[0].get_volume() << " - "
    << droplets[n_part - 1].get_volume() << " m^3" << endl;
    cout << "    mass: " << droplets[0].get_mass() << " - "
    << droplets[n_part - 1].get_mass() << " kg" << endl;

    cout << "SD SETUP" << endl;
    cout << "   N_s: " << n_part << endl;
    cout << "  xi_i: " << xi_i << endl;
    long N_per_SD = total_droplets / xi_i / n_part;
    cout << " N per SD_xi: " << N_per_SD << endl;
    cout << "(Initialized " << Droplet::global_droplet_count() << " droplets)"
         << endl;

    // ----------------------------------------------------

    cout << "\nBEGINNING MAIN ROUTINE\n" << endl;

    double wm0 = total_water(droplets, n_part);
    cout << "Initial water mass = " << wm0 << " kg" << endl;

    double t = 0.;
    int ti = 0;

    while (t < t_end) {

        // Get/print timestep info
        auto min_sec = s_to_min_s(t);
        int minutes = std::get<0>(min_sec);
        int seconds = std::get<1>(min_sec);

        cout << "\nSTEP " << ti
             << " (" << minutes << " min";
        if (seconds > 0)
            cout << " " << seconds << " sec";
        cout << ")" << endl;

        // --------------------------------
        collision_step(droplets, n_part, t_c, delta_V);
        // --------------------------------

        ti++;
        t = ti*t_c;

        cout <<  Droplet::global_droplet_count() << " droplets remain" << endl;


        if (DEBUG) break;
    }

}

File: sce.f90

program sce

    use common
    use droplet_class, only: droplet, droplet_, droplet_count

    use fgsl

    implicit none

    real(kind=rkind), parameter ::    &
        n_0 = 2**23,                  &
        R_0 = 30.531d-6,              &
        X_0 = (4d0*PI/3d0)*(R_0**3.), &
        M_0 = X_0*RHO_WATER,          &
        delta_V = 1d6,                &
        t_c = 1.0

    integer(kind=ikind), parameter :: &
        n_part = 2**11,               &
        t_end = 3601,                 &
        plot_dt = 1200

    type(droplet), dimension(n_part) :: droplets
    real(kind=rkind), dimension(n_part) :: x_grid, r_grid, m_grid


    real(kind=rkind) :: total_droplets
    real(kind=rkind) :: rcubed_i ! Placeholder for cubed radius of a droplet
    type(droplet) :: droplet_i
    integer(kind=lkind) :: xi_i

    type(fgsl_rng) :: r      ! FGSL random number generator
    type(fgsl_rng_type) :: t ! Type of FGSL random number generaotr
    real(fgsl_double) :: phi ! Holder for a sample from the RNG

    integer(kind=lkind) :: N_per_SD

    integer :: i
! -----------------------------------------------------------------------------

    print *, "BEGINNING MAIN ROUTINE"

    ! Initialize the droplet array; note that we use a statically sized
    ! array since we know the number of droplets at compile time.

    ! Pre-compute the superdroplet multiplicity
    total_droplets = delta_V * n_0
    xi_i = floor(total_droplets / n_part, lkind)

    ! 1) Initialize the random number generator from FGSL

    t = fgsl_rng_env_setup()
    t = fgsl_rng_default
    r = fgsl_rng_alloc(t)

    ! Call the random number generator to compute masses.
    do i = 1, n_part
        ! The exponential dist is of the form
        !             1
        ! p(x) dx = ---- exp ( -x / mu ) dx
        !            mu
        ! where mu = 1/lambda from the Boost version, but is the same
        ! as the scipy form of the distribution!=
        phi = fgsl_ran_exponential(r, X_0) ! phi is a sample from the mass
                                           ! distribution
        x_grid(i) = phi
    end do

    ! Pre-sort the x_grid because it's a bit easier to sort doubles
    ! with FGSL than objects
    call fgsl_sort(x_grid, 1_fgsl_size_t, &
                           int(n_part, kind=fgsl_size_t))

    r_grid = (x_grid * 3. / PI / 4.)**THIRD

    ! Populate the initial droplet array
    do i = 1, n_part
        droplets(i) = droplet_(xi_i, r_grid(i)**3.)
        m_grid(i) = x_grid(i) * droplets(i)%density
    end do

    print *, "GRID SETUP"
    print *, "   radii: ", droplets(1)%radius, " - ", &
                           droplets(n_part-1)%radius, " m"
    print *, "  volume: ", droplets(1)%volume, " - ", &
                           droplets(n_part-1)%volume, " m^3"
    print *, "    mass: ", droplets(1)%mass, " - ", &
                           droplets(n_part-1)%mass, " kg"

    print *, "SD SETUP"
    print *, "   N_s: ", n_part
    print *, "  xi_i: ", xi_i
    N_per_SD = total_droplets / xi_i / n_part
    print *, " N per SD_xi":, N_per_SD
    print *, "(Initialized ", droplet_count, " droplets)"


    ! ) Free the memory from the random number generator
    call fgsl_rng_free(r)


end program sce